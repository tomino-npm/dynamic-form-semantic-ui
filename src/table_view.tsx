import * as React from 'react';

import { observer } from 'mobx-react';
import { Button, Form } from 'semantic-ui-react';
import { FormElement, DataSet } from '@tomino/dynamic-form';

import { ErrorView } from './error_view';
import { css, FormControlProps } from './common';
import { renderControl } from './form_control_factory';

const formGroup = css`
  margin-bottom: 0px !important;
`;

type RowProps = FormControlProps & {
  data: DataSet;
  index: number;
};

class TableRow extends React.PureComponent<RowProps> {
  handlers = {
    delete: (_e: any, { value }: any) => {
      this.props.data.removeRowIndex(this.props.formControl.source, value);
    }
  };
  render() {
    return (
      <Form.Group>
        {this.props.formControl.elements.map((e, i) => (
          <Form.Field width={e.width as any} key={i}>
            {renderControl(e, this.props.owner, this.props.handlers as any, this.props.readOnly)}
          </Form.Field>
        ))}
        {!this.props.readOnly && (
          <Form.Field width={1}>
            <Button
              color="red"
              icon="trash"
              value={this.props.index}
              onClick={this.handlers.delete}
            />
          </Form.Field>
        )}
      </Form.Group>
    );
  }
}

export class TableComponent extends React.Component<FormControlProps> {
  addRow = () => {
    const {
      formControl: { source },
      owner
    } = this.props;
    owner.addRow(source);
  };

  render() {
    const owner = this.props.owner;
    const formControl = this.props.formControl as FormElement;
    const source = formControl.source;
    const list: DataSet[] = owner.getValue(source);

    return (
      <div className="ui form">
        {list && list.length && (
          <Form.Group className={formGroup}>
            {formControl.elements.map((e, i) => (
              <Form.Field key={i} label={e.label} width={e.width as any} />
            ))}
          </Form.Group>
        )}
        {list.map((row, i) => (
          <TableRow
            key={i}
            index={i}
            formControl={formControl}
            owner={row}
            data={owner}
            handlers={this.props.handlers}
            readOnly={this.props.readOnly}
          />
        ))}

        {!this.props.readOnly && (
          <Button primary icon="plus" content="Add" labelPosition="left" onClick={this.addRow} />
        )}

        <ErrorView
          owner={this.props.owner}
          source={this.props.formControl.source}
          pointing="left"
        />
      </div>
    );
  }
}

export const TableView = observer(TableComponent);
