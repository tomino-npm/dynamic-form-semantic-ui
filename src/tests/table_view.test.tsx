import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { Button, Segment } from 'semantic-ui-react';
import { JSONSchema, FormDefinition, undoManager, FormModel, config } from '@tomino/dynamic-form';

import { TableView } from '../table_view';
import { create } from './form_query_data';

describe('Form', () => {
  const schema: JSONSchema = {
    type: 'object',
    properties: {
      countries: {
        type: 'array',
        items: {
          type: 'object',
          properties: {
            id: {
              type: 'id'
            },
            name: {
              type: 'string'
            },
            capital: {
              type: 'string'
            }
          },
          required: ['name']
        }
      }
    }
  };

  const controlData = {
    countries: [
      { name: 'Slovakia', capital: 'Bratislava' },
      { name: 'Australia', capital: 'Canberra' }
    ]
  };

  const formDefinition: FormDefinition = create.form({
    elements: [
      create.formElement({
        row: 0,
        column: 0,
        width: 16,
        control: 'Table',
        source: 'countries',
        elements: [
          create.formElement({
            width: 8,
            control: 'Input',
            label: 'Name',
            source: 'name'
          }),
          create.formElement({
            width: 8,
            control: 'Input',
            label: 'Capital',
            source: 'capital'
          })
          // create.formElement({
          //   id: '4',
          //   parentElement: '1',
          //   row: 0,
          //   column: 15,
          //   width: 1,
          //   control: QueryTypes.FormControl.DeleteButton,
          //   label: '\xa0'
          // })
        ]
      })
    ]
  });

  describe('Table', () => {
    function componentWithData() {
      config.setDirty = () => console.log('Now it dirty');
      const form = new FormModel(formDefinition, schema, controlData);

      // just another notation
      return (
        <Segment className="ui form">
          <TableView formControl={form.elements[0]} owner={form.dataSet} handlers={{}} />

          <div style={{ marginTop: '20px' }}>
            <Button content="Undo" onClick={() => undoManager.undo()} />
            <Button content="Redo" onClick={() => undoManager.redo()} />
          </div>
        </Segment>
      );
    }

    it('renders correctly', () => {
      const component = renderer.create(componentWithData());
      expect(component).toMatchSnapshot();
    });

    it('changes value and all related formulas', () => {
      // const component = renderer.create(data.component);
      // const root = component.root;
      // const agree = root.findByProps({ name: 'agree' });
      // agree.props.onChange(null, { value: false });
      // const disagree = root.findByProps({ name: 'disagree' });
      // disagree.props.onChange(null, { value: true });
      // expect(component).toMatchSnapshot();
    });

    return {
      componentWithData
    };
  });
});
