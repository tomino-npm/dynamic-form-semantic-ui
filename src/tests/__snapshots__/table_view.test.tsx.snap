// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Table renders correctly 1`] = `
<div
  className="ui segment ui form"
>
  <div
    className="ui form"
  >
    <div
      className="fields d-form-Abdgbhfjheg"
    >
      <div
        className="eight wide field"
      >
        <label>
          Name
        </label>
      </div>
      <div
        className="eight wide field"
      >
        <label>
          Capital
        </label>
      </div>
    </div>
    <div
      className="fields"
    >
      <div
        className="eight wide field"
      >
        <div
          className="ui input"
        >
          <input
            name="name"
            onChange={[Function]}
            type="text"
            value="Slovakia"
          />
        </div>
      </div>
      <div
        className="eight wide field"
      >
        <div
          className="ui input"
        >
          <input
            name="capital"
            onChange={[Function]}
            type="text"
            value="Bratislava"
          />
        </div>
      </div>
      <div
        className="one wide field"
      >
        <button
          className="ui red icon button"
          onClick={[Function]}
          role="button"
          value={0}
        >
          <i
            aria-hidden="true"
            className="trash icon"
          />
        </button>
      </div>
    </div>
    <div
      className="fields"
    >
      <div
        className="eight wide field"
      >
        <div
          className="ui input"
        >
          <input
            name="name"
            onChange={[Function]}
            type="text"
            value="Australia"
          />
        </div>
      </div>
      <div
        className="eight wide field"
      >
        <div
          className="ui input"
        >
          <input
            name="capital"
            onChange={[Function]}
            type="text"
            value="Canberra"
          />
        </div>
      </div>
      <div
        className="one wide field"
      >
        <button
          className="ui red icon button"
          onClick={[Function]}
          role="button"
          value={1}
        >
          <i
            aria-hidden="true"
            className="trash icon"
          />
        </button>
      </div>
    </div>
    <button
      className="ui icon primary left labeled button"
      onClick={[Function]}
      role="button"
    >
      <i
        aria-hidden="true"
        className="plus icon"
      />
      Add
    </button>
  </div>
  <button
    className="ui button"
    onClick={[Function]}
    role="button"
  >
    Undo
  </button>
  <button
    className="ui button"
    onClick={[Function]}
    role="button"
  >
    Redo
  </button>
</div>
`;
